File: TP03/TP03Q01 - Lista com Alocação Sequencial em Java/ListaSequencial.java
/**
 * TP03Q01 - Lista Sequencial em java
 * @author: Vitória Símil Araújo  
* */

import java.io.BufferedReader;
import java.io.FileReader;

class Jogador {
    // Atributos - dados
    private int id, altura, peso, anoNascimento;
    private String nome, universidade, cidadeNascimento, estadoNascimento;

    // Construtor vazio
    public Jogador() {
        this.id = this.altura = this.peso = this.anoNascimento = 0;
        this.nome = this.universidade = this.cidadeNascimento = this.estadoNascimento = "";
    }

    /**
     * Construtor para inicializar um objeto Jogador com informações específicas.
     *
     * @param id               
     * @param altura           
     * @param peso             
     * @param anoNascimento    
     * @param nome             
     * @param universidade     
     * @param cidadeNascimento 
     * @param estadoNascimento 
     */
    public Jogador(int id, int altura, int peso, int anoNascimento, String nome, String universidade, String cidadeNascimento, String estadoNascimento) {
        this.id = id;
        this.altura = altura;
        this.peso = peso;
        this.anoNascimento = anoNascimento;
        this.nome = nome;
        this.universidade = universidade;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;
    }

    /**
    * GETTERS: @return - atributo do jogador a ser retornado
    * SETTERS: @param paramentro - novo atributo a ser atribuido ao jogador
    */

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getAltura() {
        return altura;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public int getPeso() {
        return peso;
    }

    public void setPeso(int peso) {
        this.peso = peso;
    }

    public int getAnoNascimento() {
        return anoNascimento;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getUniversidade() {
        return universidade;
    }

    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }
    
    public String getCidadeNascimento() {
        return cidadeNascimento;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }


    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    /**
     * Cria uma cópia do jogador.
     *
     * @param jogador Jogador a ser clonado.
     * @return Uma nova instância de Jogador que é uma cópia do jogador fornecido.
     */
    public Jogador clone(Jogador jogador) {
        return new Jogador(
            jogador.id,
            jogador.altura,
            jogador.peso,
            jogador.anoNascimento,
            jogador.nome,
            jogador.universidade,
            jogador.cidadeNascimento,
            jogador.estadoNascimento
        );
    }

    /**
     * Lê dados de um arquivo CSV de jogadores e os armazena em um array de strings.
     *
     * @return Um array de strings contendo os dados dos jogadores lidos do arquivo CSV.
     */
    public String[] ler() {
        // Caminho do arquivo CSV a ser lido.
        String arquivoCSV = "/tmp/players.csv";

        // Objeto BufferedReader para ler o arquivo.
        BufferedReader br = null;

        //dados
        String linha = "";
        String[] jogadores = new String[3922];
        int contador = 0;

        try {
            br = new BufferedReader(new FileReader(arquivoCSV));

            // Lê e ignora a primeira linha (cabeçalho)
            linha = br.readLine();

            // Loop para ler todas as linhas de dados dos jogadores.
            while ((linha = br.readLine()) != null) {
                jogadores[contador++] = linha;
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (br != null) {
                try {
                    // Fecha o BufferedReader.
                    br.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        return jogadores;
    }


    /**
     * Retorna uma representação em formato de string dos dados do jogador.
     *
     * @return Uma string formatada com os dados do jogador.
     */
    public String imprimir() {
        return  nome + " ## "
                + altura + " ## "
                + peso + " ## "
                + anoNascimento + " ## "
                + universidade + " ## "
                + cidadeNascimento
                + " ## " + estadoNascimento;
    }
}

/**
 * Classe que representa uma lista de jogadores.
 */
class Lista {
    // Atributos

    /** Array que armazena os jogadores. */
    public Jogador[] array;

    /** Número atual de elementos na lista. */
    public int n;

    /**
     * Construtor padrão que inicializa a lista com capacidade para 100 elementos.
     */
    public Lista() {
        this.array = new Jogador[100];
        n = 0;
    }

    /**
     * Construtor que permite especificar a capacidade inicial da lista.
     *
     * @param tamanho Capacidade inicial da lista.
     */
    public Lista(int tamanho) {
        this.array = new Jogador[tamanho];
        n = 0;
    }

    /**
     * Insere um jogador no início da lista.
     *
     * @param jogador jogador a ser inserido.
     * @throws Exception Se a lista estiver cheia.
     */
    public void inserirNoInicio(Jogador jogador) throws Exception {
        if (n == array.length) throw new Exception("Lista cheia");

        for (int i = n; i > 0; i--) {
            array[i] = array[i - 1];
        }

        array[0] = jogador;
        n++;
    }

    /**
     * Insere um jogador no final da lista.
     *
     * @param jogador jogador a ser inserido.
     * @throws Exception Se a lista estiver cheia.
     */
    public void inserirNoFim(Jogador jogador) throws Exception {
        if (n == array.length) throw new Exception("Lista cheia");

        array[n] = jogador;
        n++;
    }

    /**
     * Insere um jogador em uma posição específica da lista.
     *
     * @param pos Posição em que o jogador será inserido.
     * @param jogador jogador a ser inserido.
     * @throws Exception Se a lista estiver cheia ou a posição for inválida.
     */
    public void inserir(int pos, Jogador jogador) throws Exception {
        if (n == array.length || pos < 0 || pos > n) throw new Exception("Lista cheia ou posição inválida");

        for (int i = n; i > pos; i--) {
            array[i] = array[i - 1];
        }

        array[pos] = jogador;
        n++;
    }

    /**
     * Remove o jogador do início da lista.
     *
     * @return jogador removido.
     * @throws Exception Se a lista estiver vazia.
     */
    public Jogador removerInicio() throws Exception {
        if (n == 0) throw new Exception("Lista vazia");

        Jogador jogador = array[0];

        for (int i = 0; i < n - 1; i++) {
            array[i] = array[i + 1];
        }

        n--;

        return jogador;
    }

    /**
     * Remove o jogador do final da lista.
     *
     * @return jogador removido.
     * @throws Exception Se a lista estiver vazia.
     */
    public Jogador removerFim() throws Exception {
        if (n == 0) throw new Exception("Lista vazia");

        return array[--n];
    }

    /**
     * Remove o jogador de uma posição específica da lista.
     *
     * @param pos Posição do jogador a ser removido.
     * @return jogador removido.
     * @throws Exception Se a lista estiver vazia ou a posição for inválida.
     */
    public Jogador remover(int pos) throws Exception {
        if (n == 0 || pos < 0 || pos > n) throw new Exception("Lista vazia ou posição inválida");

        Jogador jogador = array[pos];

        for (int i = pos; i < n; i++) {
            array[i] = array[i + 1];
        }

        n--;

        return jogador;
    }

    public void mostrar() {
        for(int i = 0; i < n; i++) {
            System.out.println("["+ i+ "] ## " + array[i].imprimir() + " ##");
        }

        
    }
}


public class ListaSequencial {
    /**
     * @param entrada
     * @return boolean 
     * metodo para verificar se uma entrada é igual a FIM
     */
    private static boolean isFIM(String entrada) {
        return (entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M');
    }

    /**
     * Cria um objeto Jogador com base em uma string de dados.
     *
     * @param j  string contendo os dados do jogador no formato "id,nome,altura,peso,universidade,anoNascimento,cidadeNascimento,estadoNascimento".
     * @param jogador objeto Jogador que será preenchido com os dados extraídos da string.
     * @return objeto Jogador preenchido com os dados da string.
     */
    public static Jogador criarJogador(String j, Jogador jogador) {
        // array contendo os valores de j separado por virgula
        String[] atributos = j.split(",");

        int id = Integer.parseInt(atributos[0]);
        jogador.setId(id);

        String nome = atributos[1];
        jogador.setNome(nome);

        int altura = Integer.parseInt(atributos[2]);
        jogador.setAltura(altura);
        
        int peso = Integer.parseInt(atributos[3]);
        jogador.setPeso(peso);

        String universidade = (atributos[4].isEmpty()) ? "nao informado" : atributos[4];
        jogador.setUniversidade(universidade);

        int anoNascimento = Integer.parseInt(atributos[5]);
        jogador.setAnoNascimento(anoNascimento);
        
        String cidadeNascimento = (atributos.length == 6 ) ? "nao informado" : atributos[6];
        jogador.setCidadeNascimento(cidadeNascimento);

        String estadoNascimento = (atributos.length == 8 ) ? atributos[7]  : "nao informado";
        jogador.setEstadoNascimento(estadoNascimento);

        return jogador;
    }

    public static void main(String[] args) throws Exception{
        // Array para armazenar as linhas de dados lidas do arquivo CSV.
        String entradas[];

        Lista lista = new Lista(200);

        // Objeto Jogador para criar jogadores a partir dos dados.
        Jogador jogador = new Jogador();
        String entrada = "";

        entradas = jogador.ler();

        while (!(isFIM(entrada = MyIO.readLine()))) {
            int id = Integer.parseInt(entrada);
            jogador = criarJogador(entradas[id], jogador);
            lista.inserirNoFim(jogador.clone(jogador));
        }

        int n = Integer.parseInt(MyIO.readLine());
        int id = 0;

        while ( n > 0 ) {
            entrada = MyIO.readLine();

            String operacao = entrada.substring(0, 2);

            if ( entrada.charAt(0) == 'I') {
                id = ( entrada.charAt(1) != '*' ) ? Integer.parseInt(entrada.substring(3)) : Integer.parseInt(entrada.substring(6));

                jogador = criarJogador(entradas[id], jogador);
            }

            switch (operacao) {
                case "II":
                    lista.inserirNoInicio(jogador.clone(jogador));
                    break;

                case "IF":
                    lista.inserirNoFim(jogador.clone(jogador));
                    break;

                case "I*":
                    int pos = Integer.parseInt(entrada.substring(3,5));
        
                    lista.inserir(pos, jogador.clone(jogador));
                    break;

                case "RI":
                case "RF":
                    jogador = (operacao.charAt(1) == 'I') ? lista.removerInicio(): lista.removerFim();
        
                    System.out.println("(R) " + jogador.getNome());
        
                    break;
                
                case "R*":
                    pos = Integer.parseInt(entrada.substring(3,5));
                    
                    jogador = lista.remover(pos);
    
                    System.out.println("(R) " + jogador.getNome());
                    break;


                default:
                    break;
            }

            n--;
        }

        lista.mostrar();
    }
}

File: TP03/TP03Q02 - Lista com Alocação Sequencial em C/ListaSequencial.c
/**
 * TP03Q02 - Lista Sequencial em c
 * @author: Vitória Símil Araújo  
* */

#include "stdio.h"
#include "stdlib.h"
#include "stdbool.h"
#include "string.h"

// Definição de constantes
#define tamanhoAtributo 100
#define tamanhoEntrada 500
#define MAXTAM 200

// Declaração de uma variável global - array de entradas do arquivo CSV
char **entradas;

// Definição de uma estrutura de dados para representar um jogador
typedef struct {
    char nome[tamanhoAtributo],
        universidade[tamanhoAtributo],
        cidadeNascimento[tamanhoAtributo],
        estadoNascimento[tamanhoAtributo];

    int id, altura, peso, anoNascimento;
} Jogador;

// Definição de uma estrutura de dados para representar uma lista de jogadores
typedef struct {
    Jogador array[MAXTAM];
    int n; // Número atual de elementos na lista
} Lista;

// Cabeçalho de funções
bool isFIM(char input[]);
void ler();

//funções jogador
void criarJogador(char *entradasJogador, Jogador *jogador);
void substituirNaoInformado(char *str);
Jogador clone(Jogador *jogador);
void imprimir(Jogador *jogador);

//funções lista
void iniciar(Lista *lista);
void inserirInicio(Jogador jogador, Lista *lista);
void inserirFim(Jogador jogador, Lista *lista);
void inserir(Jogador jogador, int pos, Lista *lista);
Jogador removerInicio(Lista *lista);
Jogador removerFim(Lista *lista);
Jogador remover(int pos, Lista *lista);
void mostrar(Lista *lista);

// Função principal
int main(void) {
    // Declaração de variáveis
    Jogador jogador;
    Jogador jogadores[3922];
    char *entrada = malloc(sizeof(char*) * 10);
    
    int idJogador = 0;

    // Ler o documento CSV e armazenar os dados em um array
    ler();

    // Criar uma nova lista
    Lista lista;
    iniciar(&lista);

    // Preencher o array de jogadores com dados do arquivo CSV
    for(int i = 0; i < 3922; i++ ) {
        criarJogador(entradas[i], &jogador);
        jogadores[i] = jogador;
    }

    // Ler IDs de jogadores a partir da entrada padrão e inseri-los no final da lista
    scanf("%s", entrada);
    while(!isFIM(entrada)) {
        idJogador = atoi(entrada);
        inserirFim(clone(&jogadores[idJogador]), &lista);
        scanf("%s", entrada);
    }

    // Realizar operações adicionais na lista com base na entrada do usuário
    int n = 0, pos = 0;
    scanf("%d", &n);

    while ( n > 0 ) {
        scanf("%s ", entrada);
        if( strcmp(entrada, "II") == 0 ) {
            scanf("%d", &idJogador);
            inserirInicio(clone(&jogadores[idJogador]), &lista);
        }
        else if ( strcmp(entrada, "IF" ) == 0 ) {
            scanf("%d", &idJogador);
            inserirFim(clone(&jogadores[idJogador]), &lista);
        }
        else if ( strcmp(entrada, "I*" ) == 0 ) {
           scanf("%d", &pos);
           scanf("%d", &idJogador);
           inserir(clone(&jogadores[idJogador]), pos, &lista);
        }
        else if ( strcmp(entrada, "R*") == 0 ) {
            scanf("%d", &pos);
            jogador = remover(pos, &lista);
            printf("(R) %s\n", jogador.nome);
        }
        else if ( strcmp(entrada, "RI") == 0 ) {
            jogador = removerInicio(&lista);
            printf("(R) %s\n", jogador.nome);
        }
        else if ( strcmp(entrada, "RF" ) == 0 ){
            jogador = removerFim(&lista);
            printf("(R) %s\n", jogador.nome);
        }

        n--;
    }

    // Exibir a lista resultante
    mostrar(&lista);

    // Liberar memória alocada
    free(entrada);
    free(entradas);
}

// Função para verificar se a entrada é "FIM"
bool isFIM(char input[]) {
    return (input[0] == 'F' && input[1] == 'I' && input[2] == 'M');
}

// Função para ler os dados do arquivo CSV e armazená-los em um array
void ler() {
    char jogador[tamanhoEntrada];

    // Alocar vetor de entradas
    entradas = (char **) malloc(4000 * sizeof(char *));

    // Abrir arquivo CSV para leitura
    FILE *csv = fopen("/tmp/players.csv", "r");

    // Ler e ignorar o cabeçalho do arquivo CSV
    fgets(jogador, sizeof(jogador), csv);

    for(int i = 0; i < 3922; i++) {
        // Alocar espaço para string no array de entradas
        entradas[i] = (char *) malloc(tamanhoEntrada * sizeof(char));

        if(fgets(jogador, sizeof(jogador), csv) != NULL) {
            jogador[strcspn(jogador, "\n")] = '\0'; 
            strcpy(entradas[i], jogador);
        }
    }

    fclose(csv);
}

// Função para criar um objeto Jogador a partir de uma string de entrada CSV
void criarJogador(char *entradasJogador, Jogador *jogador) {
    substituirNaoInformado(entradasJogador);

    char *atributo;
    atributo = strtok(entradasJogador, ",");
    jogador->id = atoi(atributo);

    atributo = strtok(NULL, ",");
    strcpy(jogador->nome, atributo);
    
    atributo = strtok(NULL, ",");
    jogador->altura = atoi(atributo);

    atributo = strtok(NULL, ",");
    jogador->peso = atoi(atributo);

    atributo = strtok(NULL, ",");
    strcpy(jogador->universidade, atributo);

    atributo = strtok(NULL, ",");
    jogador->anoNascimento = atoi(atributo);

    atributo = strtok(NULL, ",");
    if(atributo == NULL) strcpy(jogador->cidadeNascimento, "nao informado");
    else strcpy(jogador->cidadeNascimento, atributo);

    atributo = strtok(NULL, ",");
    if(atributo == NULL) strcpy(jogador->estadoNascimento, "nao informado");
    else strcpy(jogador->estadoNascimento, atributo);
}

// Função para substituir valores não informados na string por "nao informado"
void substituirNaoInformado(char *str) {
    char *pos = strstr(str, ",,");
    
    while (pos != NULL) {
        int indice = pos - str;
        
        // Substituir ",," por ",nao informado,"
        memmove(str + indice + strlen("nao informado"), str + indice, strlen(str) - indice);
        strncpy(str + indice + 1, "nao informado", strlen("nao informado"));
        
        // Atualizar a posição
        pos = strstr(str, ",,");
    }
}

// Função para clonar um objeto Jogador
Jogador clone(Jogador *jogador) {
    Jogador jogadorClonado;

    strcpy(jogadorClonado.nome, jogador->nome);
    strcpy(jogadorClonado.universidade, jogador->universidade);
    strcpy(jogadorClonado.cidadeNascimento, jogador->cidadeNascimento);
    strcpy(jogadorClonado.estadoNascimento, jogador->estadoNascimento);

    jogadorClonado.id = jogador->id;
    jogadorClonado.altura = jogador->altura;
    jogadorClonado.peso = jogador->peso;
    jogadorClonado.anoNascimento = jogador->anoNascimento;

    return jogadorClonado;
}

// Função para imprimir os dados de um jogador
void imprimir(Jogador *jogador) {
    printf(" ## %s ## %i ## %i ## %i ## %s ## %s ## %s ##\n",
    jogador->nome, jogador->altura, jogador->peso,jogador->anoNascimento,jogador->universidade,
    jogador->cidadeNascimento, jogador->estadoNascimento);
}

// Função para inicializar uma lista
void iniciar(Lista *lista) {
    lista->n = 0;
}

// Função para inserir um jogador no início da lista
void inserirInicio(Jogador jogador, Lista *lista) {
    if (lista -> n == MAXTAM) {
        exit(1);
    }

    for ( int i = lista -> n; i > 0; i-- ) {
        lista -> array[i] = lista -> array[i - 1];
    }

    lista -> array[0] = jogador;

    lista -> n++;
}

// Função para inserir um jogador no final da lista
void inserirFim(Jogador jogador, Lista *lista) {
    if (lista -> n == MAXTAM) {
        exit(1);
    }

    lista -> array[lista -> n++] = jogador;
}

// Função para inserir um jogador em uma posição específica na lista
void inserir(Jogador jogador, int pos, Lista *lista) {
    if (lista -> n == MAXTAM || pos < 0 || pos > lista -> n)  {
        exit(1);
    }

    for ( int i = lista -> n; i > pos; i--) {
        lista -> array[i] = lista -> array[i - 1];
    }

    lista -> array[pos] = jogador;
    lista -> n++;
}

// Função para remover o jogador no início da lista
Jogador removerInicio(Lista *lista) {
    if (lista -> n == 0) {
        exit(1);
    }

    Jogador resposta = lista -> array[0];
    lista -> n--;

    for ( int i = 0; i < lista -> n; i++) {
        lista -> array[i] = lista -> array[i + 1];
    }

    return resposta;
}

// Função para remover o jogador no final da lista
Jogador removerFim(Lista *lista) {
    if (lista -> n == 0) {
        exit(1);
    }

    lista -> n--;

    Jogador resposta = lista -> array[lista -> n];

    return resposta;
}

// Função para remover o jogador em uma posição específica na lista
Jogador remover(int pos, Lista *lista) {
    if (lista -> n == 0) {
        exit(1);
    }

    Jogador resposta = lista -> array[pos];

    lista -> n--;

    for ( int i = pos; i < lista -> n; i++) {
        lista -> array[i] = lista -> array[ i + 1 ];
    }

    return resposta;
}

// Função para exibir os jogadores na lista
void mostrar(Lista *lista) {
    for(int i = 0; i < lista->n; i++ ) {
        printf("[%i]", i);
        imprimir(&lista->array[i]);
    }
}
